### PR TITLE
CEP 48 part 1: Sequence number-based panoramas.

### PR DESCRIPTION
**Pull request by afck**
The main component of Highway's protocol state can be viewed as a directed acyclic graph (DAG) of _units_ (the most common kind of Highway message) representing a "happened before" relation. What actually matters to the algorithm is the partial order generated by that graph: Each unit `u` cites other units `v` that `u`'s creator knew about, giving rise to an arrow `u → v` in the DAG. However, `u → v → w` implies that the creator of `u` also knew about `w`. So we write `u > v` in that situation, but also `u > w`, even if `w` is only reachable via multiple arrows. The units form a partial order, and how the generating DAG is represented doesn't matter.

Except for performance: We want to send as little data as necessary over the wire to communicate the [(strict) lower closure](https://en.wikipedia.org/wiki/Upper_set#Upper_closure_and_lower_closure) of `u`, i.e. the set of all `v` with `v < u`.

Currently we send a [panorama](https://github.com/casper-network/casper-node/blob/fa10b20aa8504f60a12e02adde00dee83ee6657b/node/src/components/consensus/highway_core/state/panorama.rs#L102), a list with one entry for each validator, containing the hash of the latest unit `v` by that validator such that `v < u`. A hash is 32 bytes.

This change reduces the size of the panorama in the wire format by replacing the hash of `v` with its sequence number, which is only 8 bytes. So when I receive `u`, and `u → v`, I don't immediately know `v`'s hash, I only know that `v` is e.g. the 42nd unit of that validator.

Since faulty validators can create forks, "the 42nd unit" is not necessarily uniquely defined anymore. A large part of this PR is about double-checking and resolving that ambiguity.

This is part of [CEP 48](https://github.com/casperlabs/ceps/pull/0048), and closes https://github.com/casper-network/casper-node/issues/1481.

### CODE REVIEW

**Review by fizyk20 — APPROVED** [review]

**Review by goral09 — COMMENTED** [review]
I'm 1/3 through reviewing your PR but I've got a question.

File: node/src/components/consensus/highway_core/highway/vertex.rs [hunk 1]
```diff
@@ -28,6 +28,11 @@ where
     C: Context,
 {
     Unit(C::Hash),
+    /// The unit with its full panorama.
+    UnitWithPanorama(C::Hash),
+    /// A unit specified by sequence number and validator index, as seen by another unit, given by
+    /// hash.
+    UnitBySeqNum(u64, ValidatorIndex, C::Hash),
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:35]: We already have the skip list to find them easily — but only for honest validators, yes.

The problem with omitting the hash is the following situation:
Imagine `v < u` and I already have multiple forks by `v`'s creator in my protocol state, but the unit `u` doesn't happen to know about them: It cites `v` directly.
Now I send you `u`, and since you don't have it yet you request `v` from me — by sequence number!
Since your request doesn't contain `u`'s hash anymore I don't know what it refers to, and I have multiple forks by `v`'s creator, so I can't find the unit. We'd be stuck.

You could argue that I should just pick _any_ unit with the right sequence number. Maybe that would be better; but it would mean that I pick a random fork for every equivocator.

To avoid complicating this feature even further I'd prefer to leave it as is for now, and consider adding such an additional index separately in the future?

[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:35]: > You could argue that I should just pick any unit with the right sequence number. Maybe that would be better; but it would mean that I pick a random fork for every equivocator.

Exactly! And if the end, the panorama's control hash doesn't match I will fall back to requesting the dependencies by hash.

Also, if you know about `V`'s forks but you don't mark `V` as faulty in your latest unit `u`, isn't that an invalid behavior and you will also end up marked as `Faulty`?

File: node/src/components/consensus/highway_core/highway/vertex.rs [hunk 2]
```diff
@@ -53,11 +68,37 @@ where
...
+    ) -> Result<S::Ok, S::Error> {
+        (swunit, panorama).serialize(serializer)
+    }
+
+    pub(super) fn deserialize<'de, D: Deserializer<'de>, C: Context>(
+        deserializer: D,
+    ) -> Result<(SignedWireUnit<C>, Panorama<C>), D::Error> {
+        let (swunit, panorama) = <(SignedWireUnit<C>, Panorama<C>)>::deserialize(deserializer)?;
+        if panorama.hash() != swunit.wire_unit().panorama_hash {
+            return Err(serde::de::Error::custom("wrong panorama hash"));
```
Comment (goral09) [inline node/src/components/consensus/highway_core/highway/vertex.rs:96]: If this is a serialization error, will we be able to ban peers if they give us invalid data? This is domain knowledge and IMO it shouldn't live in (de)serialization logic.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:35]: `u` isn't necessarily my latest unit. It's not necessarily _my_ unit at all. Maybe I'm not even a validator!

I'm a peer who sent you (maybe even as a dependency of something else) `u` (which might be old, if you've just joined).
If `u` doesn't mark `v`'s creator as faulty, that's totally fine in general.
But I should be able to answer all requests for dependencies of `u` somehow.
If I give you the wrong unit with the right sequence number, that's also fine, because as you say, you'll request it by hash if it doesn't match in the end. It might just add more rounds of communication if I send you the wrong fork first and thereby force you to request the full hash-based panorama and the other fork by hash.

Anyway, maybe you're right; at least it would simplify the different kinds of dependencies. I can give it a try and make the change. Do you think it should be part of this PR?

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 2 above)
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: I'm not sure: In general, what _do_ we do if we receive data that doesn't deserialize? I think we already report the peer as bad in that case?

Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: I definitely think that invariants that can be verified without using any data can be considered deserialization errors, though. This is just a hash that is expected to match the hash of some of the message's data.

Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: That rule isn't intrinsic to the bytes/object itself though – the message with an invalid hash can still correctly deserialize to the expected structure. To me, that rule is a domain rule and (de)serialization logic shouldn't know anything about domain, that' all.

Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: OK, I'll change it then, and do the check in `pre_validate_vertex` instead.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:35]: Thanks for the more detailed explanation, I missed some nuances originally.

> It might just add more rounds of communication if I send you the wrong fork first and thereby force you to request the full hash-based panorama and the other fork by hash.

The new approach, by design, adds more rounds of communication but that's only for the case where we have faulty/non-honest validators, and we're fine with that. This particular use-case we're discussing also involves equivocations.

I would prefer if we added it to this PR, so that it marks that piece of work as complete.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 2 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: That's only my opinion (only I feel pretty strong about it) and you don't have to do it if you strongly disagree.

File: node/src/components/consensus/highway_core/highway/vertex.rs [hunk 3]
```diff
@@ -38,6 +43,15 @@ impl<C: Context> Dependency<C> {
     pub(crate) fn is_unit(&self) -> bool {
         matches!(self, Dependency::Unit(_))
     }
+    /// Returns `true` if both refer to the same vertex, even if they differ in whether they
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:47]: (Unresolving, so I don't forget. :sweat_smile: )

[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:47]: In the final version of the code it looks like the newline is there. I commented on a _commit_ originally and I don't understand why GH doesn't show _outdated_ tag here.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 2 above)
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: I only disagree weakly, so I'll do it. :smile: 

I'd _like_ to make it impossible to create an _instance_ of `Vertex` that violates the invariant, and that would require us to make it part of the deserialization logic.
However, we already instantiate by directly calling `UnitWithPanorama(unit, panorama)` in the code anyway — an enum can only enfoce its invariants if it is made private, and wrapped in some other type.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:47]: Ah, I see, thanks! Resolving again.
(It's not "outdated" because the comment is on the _next_ line, probably. Can't put a comment on a line that's not there.)

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 2 above)
Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: If we had a separate representation of each struct for inter-node communication, then the mapping between the two could return errors. The constructor of `Vertex` is the place where _I_ would put these validation steps.

Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: > The constructor of `Vertex`

That's what I mean: An enum always has canonical constructors (its variant names), so we can't add validation there unless we wrap the whole enum in another type.

Also, if you do validation of a type in all constructors, you _should_ also do the same validation in the deserialization code, otherwise you create a way to instantiate surprising invalid objects.

Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: If there's no way to deserialize directly into an enum, and there's a single constructor that includes the domain validity rules then there's no way to create an instance of an invalid object. In that setup you would need an intermediate DTO that has a some mapping to your domain types and that DTO could be invalid from the POV of the domain but it's OK – b/c DTO is not a domain _yet_.

Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: Sure, but as long as the enum isn't private (to its module) and hidden inside another data type, you can always create (outside its module) an invalid instance by directly creating a particular variant.

Comment (goral09) [reply node/src/components/consensus/highway_core/highway/vertex.rs:96]: Sure, our job is to write the best code we can and even after making it defensive as much as possible, a malicious developer tries to work around the guards we placed and hurts himself – we can't do anything about it.

File: node/src/components/consensus/highway_core/state/tests.rs [hunk 4]
```diff
@@ -188,30 +194,85 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
...
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(N, b1, c0, N))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaLength(4)), maybe_err);
+    let maybe_err = state
+        .add_unit(unit.clone(), panorama!(a0, b1, c0))
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaMismatch(ALICE)), maybe_err);
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state/tests.rs:230]: Looks like it's duplicated with lines 216-220.

File: node/src/components/consensus/highway_core/state/tests.rs [hunk 5]
```diff
@@ -188,30 +194,85 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
...
+    let maybe_err = add_unit!(state, CAROL, 0, 4u8, None; N, N, N)
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::MissingBlock), maybe_err);
+
+    // The panorama length must be the number of validators.
+    let maybe_err = add_unit!(state, CAROL, 0, 4u8, 0xC; N, N, N, N)
+        .err()
+        .map(unit_err);
+    assert_eq!(Some(UnitError::PanoramaLength(4)), maybe_err);
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state/tests.rs:268]: Isn't that the same as in https://github.com/casper-network/casper-node/pull/1550/files#diff-5edd9079fa700dab6deb47fd36ebe675615f67a4d00a3112e0cbaeac2d1a63c4R221-R225 ?

File: node/src/components/consensus/highway_core/state/tests.rs [hunk 6]
```diff
@@ -236,6 +297,84 @@ fn add_unit() -> Result<(), AddUnitError<TestContext>> {
...
+    // We need to request the full hash-based panoramas. They contain the right c0 by hash.
+    let res = new_state.compute_panorama(&b1_swunit);
+    assert_eq!(Err(Dependency::UnitWithPanorama(b1)), res);
+    let b1_panorama = state.unit(&b1).panorama.clone();
+    let c0_swunit = state.wire_unit(&c0, TEST_INSTANCE_ID).unwrap();
+    let c0_panorama = new_state.compute_panorama(&c0_swunit).unwrap();
+    new_state.add_unit(c0_swunit, c0_panorama)?;
+
+    // Finally we can add b1 and a2. Since a2 already knows that Carol is faulty, its panorama can
+    // be computed despite the multiple forks.
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state/tests.rs:370]: Can you add an assertion that verifies that `a2`'s panorama indeed sees Carol as faulty?

File: node/src/components/consensus/highway_core/state.rs [hunk 7]
```diff
@@ -768,31 +862,76 @@ impl<C: Context> State<C> {
...
         }
+        for ((vidx, obs), seq_obs) in panorama.enumerate().zip(&wunit.seq_num_panorama) {
+            match (obs, seq_obs) {
+                (Observation::Faulty, ObservationSeqNum::Faulty)
+                | (Observation::None, ObservationSeqNum::None) => {}
+                (Observation::Correct(hash), ObservationSeqNum::Correct(seq_number))
+                    if self.unit(hash).seq_number == *seq_number => {}
+                (_, _) => return Err(UnitError::PanoramaMismatch(vidx)),
+            }
+        }
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state.rs:934]: Can we extract this to a method on `Panorama` type? Also, it seems like you do that in `pre_validate_unit` as well.

File: node/src/components/consensus/highway_core/state.rs [hunk 8]
```diff
@@ -768,31 +862,76 @@ impl<C: Context> State<C> {
...
+            (ObservationSeqNum::Correct(seq_num), Some(_)) => {
+                if seq_num.checked_add(1) != Some(wunit.seq_number) {
+                    return Err(UnitError::SequenceNumber);
+                }
+            }
+            (ObservationSeqNum::None, None) if wunit.seq_number != 0 => {
+                return Err(UnitError::SequenceNumber);
+            }
+            _ => {}
+        }
```
Comment (goral09) [inline node/src/components/consensus/highway_core/state.rs:890]: Can you add a comment that we validate unit's `previous` against the panorama for consistency?

File: node/src/components/consensus/highway_core/highway.rs [hunk 9]
```diff
@@ -352,21 +419,36 @@ impl<C: Context> Highway<C> {
...
+            Dependency::UnitBySeqNum(seq_num, vidx, hash) => self
+                .validators
+                .id(*vidx)
+                .and_then(|_| self.state.maybe_unit(hash))
+                .and_then(|unit| unit.panorama[*vidx].correct())
+                .and_then(|hash| self.state.find_in_swimlane(hash, *seq_num))
+                .and_then(|hash| self.state.wire_unit(hash, self.instance_id))
+                .map_or(GetDepOutcome::None, |wunit| {
+                    GetDepOutcome::Vertex(Vertex::Unit(wunit))
+                }),
```
Comment (goral09) [inline node/src/components/consensus/highway_core/highway.rs:440]: This should get simplified if we get rid of `hash` from `UnitBySeqNum`.

**Review by goral09 — APPROVED** [review]
Looks good, I don't think I had any blocking comments.

File: node/src/components/consensus/highway_core/active_validator.rs [hunk 10]
```diff
@@ -197,16 +195,14 @@ impl<C: Context> ActiveValidator<C> {
                 return effects;
             } else if timestamp == r_id + self.witness_offset(r_len) {
                 let panorama = self.panorama_at(state, timestamp);
-                if let Some(witness_unit) =
-                    self.new_unit(panorama, timestamp, None, state, instance_id)
-                {
+                if let Some(vv) = self.new_unit(panorama, timestamp, None, state, instance_id) {
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/active_validator.rs:198]: Or `witness_vv` and `confirmation_vv`, since it's a `ValidVertex` now?

File: node/src/components/consensus/highway_core/state/panorama.rs [hunk 11]
```diff
@@ -213,6 +213,11 @@ impl<C: Context> Panorama<C> {
         self.enumerate().filter_map(missing_dep).next()
     }
 
+    /// Returns the checksum of all hashes in this panorama.
+    pub(crate) fn hash(&self) -> C::Hash {
+        <C as Context>::hash(&bincode::serialize(self).expect("serialize Panorama"))
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/state/panorama.rs:218]: Isn't that identical? The `ValidatorMap` is just a `Vec` wrapper.

[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/state/panorama.rs:218]: Isn't the wrapper adding to the serialization format? If not, then it's all good already.

File: node/src/components/consensus/highway_core/state/tests.rs (hunk 4 above)
Comment (afck) [reply node/src/components/consensus/highway_core/state/tests.rs:230]: Good catch! :+1:

File: node/src/components/consensus/highway_core/state/tests.rs (hunk 6 above)
Comment (afck) [reply node/src/components/consensus/highway_core/state/tests.rs:370]: But that's just the `F` in line 319:
```rust
let a2 = add_unit!(state, ALICE, None; a1, b1, F)?;
```

File: node/src/components/consensus/highway_core/state/panorama.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/state/panorama.rs:218]: I'll double check, but I'm pretty sure it never does.

File: node/src/components/consensus/highway_core/active_validator.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (goral09) [reply node/src/components/consensus/highway_core/active_validator.rs:198]: Ok, sounds good.

File: node/src/components/consensus/highway_core/state.rs (hunk 7 above)
Comment (afck) [reply node/src/components/consensus/highway_core/state.rs:934]: It's not exactly the same, though, because in `validate_unit` we compare sequence numbers, and in `pre_validate_unit` we can't (because we don't necessarily have the unit with the cited hash yet).

File: node/src/components/consensus/highway_core/highway.rs [hunk 12]
```diff
@@ -280,7 +295,14 @@ impl<C: Context> Highway<C> {
     ) -> Result<ValidVertex<C>, (PreValidatedVertex<C>, VertexError)> {
         match self.do_validate_vertex(pvv.inner()) {
             Err(err) => Err((pvv, err)),
-            Ok(()) => Ok(ValidVertex(pvv.0)),
+            Ok(Some(panorama)) => match pvv.0 {
+                Vertex::Unit(swunit) => Ok(ValidVertex(Vertex::UnitWithPanorama(swunit, panorama))),
+                vertex => {
+                    error!(?vertex, "unexpected computed panorama for vertex");
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway.rs:302]: You're right; we actually don't need the separate `do_validate_vertex` at all anymore, then this problem goes away, too.

File: node/src/components/consensus/highway_core/highway.rs [hunk 13]
```diff
@@ -373,23 +373,41 @@ impl<C: Context> Highway<C> {
     }
 
     /// Returns whether we have a vertex that satisfies the dependency.
-    pub(crate) fn has_dependency(&self, dependency: &Dependency<C>) -> bool {
+    ///
+    /// If it's ambiguous, i.e. the dependency is specified by sequence number and we don't have
+    /// the citing unit, `None` is returned.
```
[the remark this answers is not in the document]
Comment (afck) [reply node/src/components/consensus/highway_core/highway.rs:378]: It actually should always return `None` or `Some(false)` for `UnitBySeqNum` now, because we can _never_ know whether that's the right unit.

**Comment by afck** [comment]
Thanks, @goral09! I addressed or replied to all your comments. Please take another look.

**Review by goral09 — APPROVED** [review]
Thanks for the updates @afck. Looks like this https://github.com/casper-network/casper-node/pull/1550#discussion_r649991739 is the only one outstanding but that's not a blocker. Still approve :heavy_check_mark:

**Review by xcthulhu — CHANGES_REQUESTED** [review]
Since these changes are substantial, it would be best to extend the DES tests in `highway_core` to cover this code and various corner cases to avoid a regression.

A less attractive alternative is to do gray-box testing with the `test_chain` rig.

File: node/src/components/consensus/highway_core/highway/vertex.rs (hunk 2 above)
Comment (afck) [reply node/src/components/consensus/highway_core/highway/vertex.rs]: Done in https://github.com/casper-network/casper-node/pull/1550/commits/bda70086de3433cf17a4a8341138fce929fafd12.

**Comment by goral09** [comment]
> Since these changes are substantial, it would be best to extend the DES tests in `highway_core` to cover this code and various corner cases to avoid a regression.
> 
> A less attractive alternative is to do gray-box testing with the `test_chain` rig.

This PR is being merged to a feature branch, not `dev`. DES tests in `highway_core` can't be extended to test the synchronizer b/c it [synchronizer] is not part of `highway_core` in a sense that can be tested there.
